### ISSUE TITLE
Code Review Feedback

### ISSUE THREAD

**Issue by bundacia** [jrpool/json-db-cli#1]
### File Organization
* It would probably make more sense for `commands` and `task.js` to be inside the `src` directory.
* Great idea to move the text of the messages out into a separate file!
* Kudos for having a separate file to do the command line parsing and validation and then delegating to command functions stored in separate files!

### Use of Comments

I've noticed you use a lot of comments. Nearly 1 per line in some files. This can be helpful as you're learning, but these types of comments really get in the way in an active codebase. There are a few reason lots of comments are widely considered a "code smell". First, your code should be clear enough that someone reading it later (you or someone else) should be able to understand it without comments most of the time. If you have to use a comment to explain what a piece of code is doing, then rewrite that code until it's obvious what it's doing without a comment. Second, comments have to be maintained just like real code. When you come back to change code that's littered with lots of comments and need to make changes you now have two jobs. Update the code to implement the new feature, and update all the comments to reflect the changes in the code. The problem is that it's really easy to forget to do the second part, or just miss a comment or two. That's because your tests won't tell you if your comments are correct, just your code. So it's really easy for comments to start to lie over time, stating things as fact that are no longer true and causing a lot of confusion. 
## Commands
* The code to build the file path for the json file and read it, parse the json, and handle errors is duplicated in all of these files. Consider refactoring that piece out into a shared function. Same with the code to serialize the tasks and write the file.
* Consider using an array to store the tasks instead of an object. It's a more natural fit for a list like this and would make some of the code simpler.
* Consider renaming `resultAction` to `callback`. That's probably more idiomatic in javascript.
* In `list.js` this code could be simplified:

```javascript
 // Sort them.
 itemIDs.sort((a, b) => a - b);
 // For each of them, add a row to the table.
 itemIDs.forEach(currentValue => {
   table.push([currentValue, listObject[currentValue]]);
 });
```

to this:

```javascript
itemIDs.sort();
const rows = itemIDs.map(id => [id, listObject[id]])
table.push(...rows);
```

### Test Isolation and State 
An important rule to follow when writing tests is that each individual test should stand alone and not permanently change any external state of depend on some state change made by another test. This helps you avoid a lot of potential bugs in your tests by keeping them simple and isolated. It also allows you to run a single test in isolation (which will save you a lot of time when you have a test suite that takes more than a few seconds to run). In your tests you are setting a global variable in the first test (`thisId`) that is used in latter tests. That's a clever solution for the problem you're solving but there are better ways. Try using a mocha `beforeEach()` block to reset the tasks.json file to a known state before each test so that you'll always know when the tests run exactly what the task list looks like.

### Use of chai `expect`
The way you're using the chai `expect` assertions is a little harder to read than it needs to be. For example:
```javascript
expect(/^Created task \d+\./.test(response)).true;
```
can be written like this:
```javascript
expect(/^Created task \d+\./.test(response)).to.be.true;
```

This is a case where the *more* verbose options actually makes the code easier to follow since it reads more like english. Similarly, `expect(response).equal(` can be `expect(response).to.equal(` and for matching a regular expression you can do this:

```javascript
expect(response).to.match(/^.+ID.+Description.+\d+ tasks?\..*\n$/);
```
instead of this:

```javascript
expect(/^.+ID.+Description.+\d+ tasks?\..*\n$/.test(response).true);
```

In fact, in that last example you've got a misplaced closing parenthesis at the end and so there's a bug in that test. The regex is actually not matching the response but because of the misplaced paren the test is not failing like it should.

**Comment by jrpool** [comment]
Thank you very much for the detailed comments. There is much to study in these. I’ll explore and see what I can learn from them.

The main thing I thought you might take exception to is my use of child_process to run system commands, emulating a user. Since you didn’t, I’ll assume that this is a reasonable approach to testing a CLI application with functions that don’t return values.

**Comment by jrpool** [comment]
I initially put task.js into the “src” directory, but found no way to satisfy the spec that “node task …” be the command, unless I made “src” the working directory. It seemed simpler not to force the user to do that. I looked for info on making node automatically look in a “src” subdirectory of process.cwd() for its file argument, but found no solution, so decided to elevate task.js to the project’s home directory. If you know of a solution that I missed, please advise. Thank you.

**Comment by bundacia** [comment]
Yes, `child_process` is a fine way to do these types integration tests. Another option would be to write tests unit for the individual command modules and call those functions directly in the tests. For a full-scale production app I'd probably do some of both, testing the specific edge cases of of the behavior of each command in unit tests that call the command functions directly and then running a few end-to-end integration tests using `child_process` like you are here. 

And you're right about the spec. It does assume that the task.js file is at the top level. If you move it into the src directory you'd have to execute the commands this way: `node src/task.js add "some task"`. Which wouldn't really be to spec.
